### DISCUSSION TITLE
Converting large zsh completion (for 1password) to xonsh

### DISCUSSION THREAD

**Discussion by drmikecrowe** [xonsh/xonsh#5582] [Q&A]
I'd love to tackle [1Password completions](https://developer.1password.com/docs/cli/reference/commands/completion) as a project.  However, the code looks pretty complex.

Here's their [zsh](https://gist.github.com/drmikecrowe/b8e6a8d77dcbc3e63d380d39af38614e) completion.  Their fish completion is similarly sized, but their bash completion is about 4k lines (I can provide if that helps the conversation).

Are there any tools to assist with building completions?  Any guidance on how to tackle this?

**Comment by anki-code** [comment]
@jnoortheen please mention the xonsh projects/xontribs with good completion as example.

**Comment by jnoortheen** [comment]
@drmikecrowe you can start looking from these places 

- https://github.com/xonsh/xonsh/tree/main/xompletions
- https://github.com/xonsh/xonsh/tree/main/xonsh/completers

Basically, the completers are just a function expected to return a set of strings/RichCompletion instances. 

Alternatively, you can build [ArgumentParser](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser) instance and pass it through https://github.com/xonsh/xonsh/blob/7c2f0f404baea6df1dad906648d394082200c83d/xonsh/cli_utils.py#L414 to get completions and positions handled automatically.
